phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/RetryingComJSEnv.scala

```diff
@@ -30,165 +30,103 @@ import scala.util.{Try, Failure, Success}
+final class RetryingComJSEnv(val baseEnv: JSEnv, val maxRetries: Int)
+    extends JSEnv {
 
-  def this(baseEnv: ComJSEnv) = this(baseEnv, 5)
+  def this(baseEnv: JSEnv) = this(baseEnv, 5)
 
-  def name: String = s"Retrying ${baseEnv.name}"
+  val name: String = s"Retrying ${baseEnv.name}"
 
-  def jsRunner(files: Seq[VirtualJSFile]): JSRunner =
-    baseEnv.jsRunner(files)
+  def start(input: Input, runConfig: RunConfig): JSRun =
+    baseEnv.start(input, runConfig)
 
-  def asyncRunner(files: Seq[VirtualJSFile]): AsyncJSRunner =
-    baseEnv.asyncRunner(files)
-
-  def comRunner(files: Seq[VirtualJSFile]): ComJSRunner =
-    new RetryingComJSRunner(files)
-
-  /** Hack to work around abstract override in ComJSRunner */
-  private trait DummyJSRunner {
-    def stop(): Unit = ()
+  def startWithCom(input: Input, runConfig: RunConfig,
+      onMessage: String => Unit): JSComRun = {
+    new RetryingJSComRun(input, runConfig, onMessage)
   }
 
-  private class RetryingComJSRunner(files: Seq[VirtualJSFile])
-      extends DummyJSRunner with ComJSRunner {
+  private class RetryingJSComRun(input: Input, runConfig: RunConfig,
+      onMessage: String => Unit)
+      extends JSComRun {
 
     private[this] val promise = Promise[Unit]
 
-    private[this] var curRunner = baseEnv.comRunner(files)
+    def future: Future[Unit] = promise.future
 
     private[this] var hasReceived = false
     private[this] var retryCount = 0
 
     private[this] val log = mutable.Buffer.empty[LogItem]
 
-    private[this] var _logger: Logger = _
-    private[this] var _console: JSConsole = _
-
-    def future: Future[Unit] = promise.future
+    private[this] var curRun: JSComRun = _
+    startNewRun()
 
-    def start(logger: Logger, console: JSConsole): Future[Unit] = {
-      require(log.isEmpty, "start() may only be called once")
+    private def startNewRun(): Unit = {
+      import ExecutionContext.Implicits.global
 
-      _logger = logger
-      _console = console
-
-      logAndDo(Start)
-      future
-    }
-
-    override def stop(): Unit = {
-      require(log.nonEmpty, "start() must have been called")
-      close()
-      logAndDo(Stop)
-    }
+      val myRetryCount = retryCount
 
-    def send(msg: String): Unit = {
-      require(log.nonEmpty, "start() must have been called")
-      logAndDo(Send(msg))
-    }
-
-    def receive(timeout: Duration): String = {
-      @tailrec
-      def recLoop(): String = {
-        // Need to use Try for tailrec
-        Try {
-          val result = curRunner.receive(timeout)
-          // At this point, we are sending state to the JVM, we cannot retry
-          // after this.
-          hasReceived = true
-          result
-        } match {
-          case Failure(t) =>
-            retry(t)
-            recLoop()
-          case Success(v) => v
+      val run = baseEnv.startWithCom(input, runConfig, { msg =>
+        val doForward = synchronized {
+          if (retryCount == myRetryCount) {
```

Comment:
This should not be needed. If the `future` of a run completes, the callback handler must not be called anymore. If it does, it is a bug.